### PR TITLE
SONARJAVA-1651 handle boolean wrapper type

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java [hunk 1]
```diff
@@ -1627,3 +1627,32 @@ public void run() {
...
+  void test1(Boolean condition) {
+    if (Boolean.FALSE.equals(condition)) {
+    } else if (Boolean.TRUE.equals(condition)) {
+    } else if (condition == null) { // Noncompliant
+    }
+  }
+
+  void test2(Boolean condition) {
+    if (Boolean.TRUE.equals(condition)) {
+    } else if (condition == null) { // Noncompliant FP
```
Comment (m-g-sonar) [inline java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java]: still a WIP.

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinarySymbolicValue.java [hunk 2]
```diff
@@ -53,12 +54,20 @@ public void computedFrom(List<SymbolicValue> symbolicValues) {
 
   protected List<ProgramState> copyConstraint(SymbolicValue from, SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinarySymbolicValue.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "copyConstraint" is 14 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinarySymbolicValue.java [hunk 3]
```diff
@@ -67,4 +76,29 @@ public void computedFrom(List<SymbolicValue> symbolicValues) {
...
+  private List<ProgramState> handleBooleanWrapperWhenNull(SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint) {
+    if (programState.getConstraint(to) == null) {
+      return ImmutableList.<ProgramState>builder()
+        .addAll(to.setConstraint(programState, BooleanConstraint.TRUE))
+        .addAll(to.setConstraint(programState, BooleanConstraint.FALSE)).build();
+    }
+    return shouldNotInverse().equals(booleanConstraint) ? ImmutableList.of(programState) : ImmutableList.of();
+  }
+
+  private List<ProgramState> handleBooleanWrapperWhenBoolean(SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint, List<ProgramState> newProgramStates) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinarySymbolicValue.java]: MINOR Split this 181 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java [hunk 4]
```diff
@@ -155,6 +156,19 @@ public SymbolicValue wrappedValue() {
     return this;
   }
 
+  public static class TypedSymbolicValue extends SymbolicValue {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java]: MAJOR Override this superclass' "equals" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2160)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MAJOR 3 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SymbolicValue.java#L115](https://github.com/SonarSource/sonar-java/blob/27435269cd13ab1f97d07282fd985fe2093239ed/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L115): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)
1. MAJOR [SymbolicValue.java#L115](https://github.com/SonarSource/sonar-java/blob/27435269cd13ab1f97d07282fd985fe2093239ed/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L115): The Cyclomatic Complexity of this method "setConstraint" is 15 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SymbolicValue.java#L115](https://github.com/SonarSource/sonar-java/blob/2abe505c147474c41a72a5c7f88cf994c20ef2c0/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L115): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)
1. MAJOR [SymbolicValue.java#L115](https://github.com/SonarSource/sonar-java/blob/2abe505c147474c41a72a5c7f88cf994c20ef2c0/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L115): The Cyclomatic Complexity of this method "setConstraint" is 15 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SymbolicValue.java#L115](https://github.com/SonarSource/sonar-java/blob/8c55e050ad6a83cc453a9464772fff6c184009bc/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L115): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)
1. MAJOR [SymbolicValue.java#L115](https://github.com/SonarSource/sonar-java/blob/8c55e050ad6a83cc453a9464772fff6c184009bc/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L115): The Cyclomatic Complexity of this method "setConstraint" is 14 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 5]
```diff
@@ -123,6 +131,16 @@ public BooleanConstraint shouldNotInverse() {
     return results;
   }
 
+  private List<ProgramState> addNullConstraintsForBooleanWrapper(BooleanConstraint booleanConstraint, ProgramState initialProgramState, List<ProgramState> copiedConstraints) {
+    Constraint leftConstraint = initialProgramState.getConstraint(leftOp);
+    Constraint rightConstraint = initialProgramState.getConstraint(rightOp);
+    if (leftConstraint instanceof BooleanConstraint && rightConstraint == null && !shouldNotInverse().equals(booleanConstraint)) {
+      List<ProgramState> nullConstraints = copiedConstraints.stream().map(ps -> ps.addConstraint(rightOp, ObjectConstraint.nullConstraint())).collect(Collectors.toList());
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)
MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)
MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MAJOR 4 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SymbolicValue.java#L114](https://github.com/SonarSource/sonar-java/blob/5a5c12b84effe2348f520dec3a27a8aeefecf66b/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L114): The Cyclomatic Complexity of this method "setConstraint" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Pull request merged**